### ISSUE TITLE
Strange non-determinism and errors when indexing NYT collection

### ISSUE THREAD

**Issue by lintool** [castorini/Anserini#234]
This is the indexing command I ran on `tuna`:

```
nohup sh target/appassembler/bin/IndexCollection -collection TrecCoreCollection \
 -input /tuna1/collections/newswire/NYTcorpus.uncompressed/data/ -generator JsoupGenerator \
 -index lucene-index.nyt.pos+docvectors+rawdocs -threads 16 \
 -storePositions -storeDocvectors -storeRawDocs -optimize \
 >& log.nyt.pos+docvectors+rawdocs &
```

This is the detail of the log:

```
$ tail log.nyt.pos+docvectors+rawdocs
2018-06-12 16:53:38,081 INFO  [main] index.IndexCollection (IndexCollection.java:277) - Indexed documents: 1807636
2018-06-12 16:53:38,082 INFO  [main] index.IndexCollection (IndexCollection.java:278) - Empty documents: 0
2018-06-12 16:53:38,082 INFO  [main] index.IndexCollection (IndexCollection.java:279) - Errors: 48022
2018-06-12 16:53:38,088 INFO  [main] index.IndexCollection (IndexCollection.java:282) - Total 1807636 documents indexed in 00:05:58
```

You can see a number of errors above. These correspond to "0 docs added" below:

```
$ grep "0 docs added" log.nyt.pos+docvectors+rawdocs | head
2018-06-12 16:47:49,499 INFO  [pool-3-thread-2] index.IndexCollection$IndexerThread (IndexCollection.java:160) - 07/0054831.xml: 0 docs added.
2018-06-12 16:47:49,499 INFO  [pool-3-thread-5] index.IndexCollection$IndexerThread (IndexCollection.java:160) - 07/0054915.xml: 0 docs added.
2018-06-12 16:47:49,954 INFO  [pool-3-thread-16] index.IndexCollection$IndexerThread (IndexCollection.java:160) - 07/0054852.xml: 0 docs added.
2018-06-12 16:47:49,967 INFO  [pool-3-thread-6] index.IndexCollection$IndexerThread (IndexCollection.java:160) - 07/0054845.xml: 0 docs added.
2018-06-12 16:47:49,975 INFO  [pool-3-thread-10] index.IndexCollection$IndexerThread (IndexCollection.java:160) - 07/0054905.xml: 0 docs added.
2018-06-12 16:47:52,411 INFO  [pool-3-thread-6] index.IndexCollection$IndexerThread (IndexCollection.java:160) - 26/0060269.xml: 0 docs added.
2018-06-12 16:47:52,456 INFO  [pool-3-thread-6] index.IndexCollection$IndexerThread (IndexCollection.java:160) - 26/0060214.xml: 0 docs added.
2018-06-12 16:47:52,489 INFO  [pool-3-thread-5] index.IndexCollection$IndexerThread (IndexCollection.java:160) - 26/0060111.xml: 0 docs added.
2018-06-12 16:47:52,496 INFO  [pool-3-thread-10] index.IndexCollection$IndexerThread (IndexCollection.java:160) - 26/0060057.xml: 0 docs added.
2018-06-12 16:47:52,650 INFO  [pool-3-thread-16] index.IndexCollection$IndexerThread (IndexCollection.java:160) - 26/0059896.xml: 0 docs added.
```

What's strange is that I run the indexer again, and I get a different number of documents:

```
2018-06-12 17:22:12,352 INFO  [main] index.IndexCollection (IndexCollection.java:277) - Indexed documents: 1809817
2018-06-12 17:22:12,352 INFO  [main] index.IndexCollection (IndexCollection.java:278) - Empty documents: 0
2018-06-12 17:22:12,352 INFO  [main] index.IndexCollection (IndexCollection.java:279) - Errors: 45841
2018-06-12 17:22:12,358 INFO  [main] index.IndexCollection (IndexCollection.java:282) - Total 1809817 documents indexed in 00:06:45
```

I suspect it's some type of multi-threaded synchronization issue, so I re-indexed, but only with four threads this time:

```
2018-06-12 17:49:24,139 INFO  [main] index.IndexCollection (IndexCollection.java:277) - Indexed documents: 1825983
2018-06-12 17:49:24,139 INFO  [main] index.IndexCollection (IndexCollection.java:278) - Empty documents: 0
2018-06-12 17:49:24,139 INFO  [main] index.IndexCollection (IndexCollection.java:279) - Errors: 29675
2018-06-12 17:49:24,146 INFO  [main] index.IndexCollection (IndexCollection.java:282) - Total 1825983 documents indexed in 00:14:09
```

Fewer errors.

The MAP/P30 are slightly different each time as a result...

**Comment by Kytabyte** [comment]
Multithreading is an issue as more threads brings more errors.

However, when I run the indexing using 1 thread, there are still errors but the number is deterministic, 24549.

```
2018-06-12 22:22:32,604 INFO  [main] index.IndexCollection (IndexCollection.java:279) - Indexed documents: 1831109
2018-06-12 22:22:32,604 INFO  [main] index.IndexCollection (IndexCollection.java:280) - Empty documents: 0
2018-06-12 22:22:32,604 INFO  [main] index.IndexCollection (IndexCollection.java:281) - Errors: 24549
2018-06-12 22:22:32,610 INFO  [main] index.IndexCollection (IndexCollection.java:284) - Total 1831109 documents indexed in 00:38:10
```

**Comment by Kytabyte** [comment]
@lintool @Peilin-Yang 

This issue comes from two problems: 1. null document not get handled before returning the document content and 2. errors while multithreaded running.

1. The body content of a document, in NYT corpus, is identified as a `BODY` tag with `full_text` attribute, and the content will only be set when someone calling

https://github.com/castorini/Anserini/blob/c81344fbe5dd2852b2cabe8249b101254ed9ed4d/src/main/java/io/anserini/document/nyt/NYTCorpusDocumentParser.java#L420

otherwise it leaves `null`.

The error comes when there is no `full_text` attribute in docs.

I run `grep -r "full_text" | wc -l` to check how many files contain the `full_text` string, and get the result
```
1831109
```
which equals the number of indexed document when I indexed the document using 1 thread
```
2018-06-12 22:22:32,604 INFO  [main] index.IndexCollection (IndexCollection.java:279) - Indexed documents: 1831109
2018-06-12 22:22:32,604 INFO  [main] index.IndexCollection (IndexCollection.java:280) - Empty documents: 0
2018-06-12 22:22:32,604 INFO  [main] index.IndexCollection (IndexCollection.java:281) - Errors: 24549
2018-06-12 22:22:32,610 INFO  [main] index.IndexCollection (IndexCollection.java:284) - Total 1831109 documents indexed in 00:38:10
```
If I changed this line
https://github.com/castorini/Anserini/blob/c81344fbe5dd2852b2cabe8249b101254ed9ed4d/src/main/java/io/anserini/document/TrecCoreDocument.java#L49
to
```
contents = nytDoc.getBody() == null ? "" : nytDoc.getBody();
```
These errors went into empty documents.


2.  The multithreaded issue is from https://github.com/castorini/Anserini/blob/c81344fbe5dd2852b2cabe8249b101254ed9ed4d/src/main/java/io/anserini/document/nyt/NYTCorpusDocumentParser.java#L68

where the `format` object obviously will be shared among all instances of `NYTCorpusDocumentParser` class, where the mutex will happen. After I remove the `static final` to make it as a object-individual variable and reindex

```
nohup sh target/appassembler/bin/IndexCollection -collection TrecCoreCollection  -input /tuna1/collections/newswire/NYTcorpus.uncompressed/data -generator JsoupGenerator  -index lucene-index.core.pos+docvectors -threads 16 -storePositions -storeDocvectors -optimize  >& log.core.pos+docvectors
```
The error becomes 24549, which is the same error number from the first bug
```
2018-06-28 12:05:37,411 INFO  [main] index.IndexCollection (IndexCollection.java:302) - Indexed documents: 1831109
2018-06-28 12:05:37,411 INFO  [main] index.IndexCollection (IndexCollection.java:303) - Empty documents: 0
2018-06-28 12:05:37,411 INFO  [main] index.IndexCollection (IndexCollection.java:304) - Errors: 24549
2018-06-28 12:05:37,419 INFO  [main] index.IndexCollection (IndexCollection.java:307) - Total 1831109 documents indexed in 00:21:47
```

So far I've only done the separate test for these two bugs, I'll get them together and do some more tests to confirm it's bug free now, and do the pull request soon.

**Comment by Kytabyte** [comment]
After fixed both bugs I re-indexed with

```
nohup sh target/appassembler/bin/IndexCollection -collection TrecCoreCollection  -input /tuna1/collections/newswire/NYTcorpus.uncompressed/data -generator JsoupGenerator  -index lucene-index.core.pos+docvectors -threads 16 -storePositions -storeDocvectors -optimize  >& log.core.pos+docvectors
```
and get results
```
2018-06-28 14:31:19,947 INFO  [main] index.IndexCollection (IndexCollection.java:303) - Indexed documents: 1831109
2018-06-28 14:31:19,948 INFO  [main] index.IndexCollection (IndexCollection.java:304) - Empty documents: 24549
2018-06-28 14:31:19,948 INFO  [main] index.IndexCollection (IndexCollection.java:305) - Errors: 0
2018-06-28 14:31:20,083 INFO  [main] index.IndexCollection (IndexCollection.java:308) - Total 1831109 documents indexed in 00:38:51
```

**Comment by lintool** [comment]
Thanks!

Next steps:
+ Please update `docs/experiments-core17.md`
+ Add regression test script, see `src/main/python/run_regression_robust04.py`

And then send PR.

**Comment by Peilin-Yang** [comment]
Thanks for digging deep into this issue, very helpful!

**Comment by lintool** [comment]
Fixed in PR #283
